runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, secondDimLenReg, generateS390MemoryReference(dimsPtrReg, 0, cg));
+
+
+   // Check to see if second dimension is indeed 0. If yes, then proceed to handle the case here. Otherwise jump to OOL code to perform the call.
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, secondDimLenReg, (int8_t)0, oolFailLabel, (TR::InstOpCode::COND_BNE));
+
+   // Now check to see if first dimension is also 0. If no, then jump to nonZeroFirstDimLabel to handle it. Otherwise continue below to handle the case when length for both dimensions is 0.
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, firstDimLenReg, (int8_t)0, nonZeroFirstDimLabel, (TR::InstOpCode::COND_BNE));
+
+   // TODO: We don't need this label. But it's good for organizing the different paths for now.
+   TR::LabelSymbol *firstDimZeroLabel = generateLabelSymbol(cg);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, firstDimZeroLabel);
+
+   // First Dim zero, only allocate 1 zero-length object array
+   int32_t zeroArraySize = TR::Compiler->om.discontiguousArrayHeaderSizeInBytes();
+   TR::Register *vmThreadReg = cg->getMethodMetaDataRealRegister();
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetReg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapAlloc), cg));
+
+   // Branch to OOL if there's not enough space for an array of size 0.
+   TR::Register *temp1Reg = cg->allocateRegister();
+   generateRIEInstruction(cg, TR::InstOpCode::getAddHalfWordImmDistinctOperandOpCode(), node, temp1Reg, targetReg, zeroArraySize);
+   generateRXInstruction(cg, TR::InstOpCode::getCmpLogicalOpCode(), node, temp1Reg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapTop), cg));
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BH, node, oolFailLabel);
+
+   // If there's enough space, then we can continue to allocate.
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp1Reg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapAlloc), cg));
+   
+   // Init class
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+   
+   // Init size and '0' fields to 0
+   TR::Register *temp0Reg = cg->allocateRegister();
+   generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, temp0Reg, temp0Reg);
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp0Reg, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg));
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp0Reg, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg));
```

Comment:
We don't need a register for this. We can use `MVHI` / `MVGHI`. Also does this even work under gencon? What would those two `fej9->get*` APIs return in gencon?

Reply:
Using move instructions now in 294933a and ff6da2b. 

I've never had any problems in any particular GC policy. 

On a compressedrefs jvm on zlinux, I get the following output for those queries:
```
fej9->getOffsetOfContiguousArraySizeField(): 4, fej9->getOffsetOfDiscontiguousArraySizeField(): 8
```